File: services/api/src/services/community/ForumService.ts
import { Op } from 'sequelize';

import ForumCategory from '../../models/community/ForumCategory';
import ForumPost from '../../models/community/ForumPost';
import ForumThread from '../../models/community/ForumThread';
import ForumVote from '../../models/community/ForumVote';
import { User } from '../../models/User';
import { logger } from '../../utils/logger';
import { getCacheService } from '../cache/UnifiedCacheService';
import DOMPurify from 'isomorphic-dompurify';
import { sequelize } from '../../config/database';

interface CreateThreadData {
  categoryId: string;
  userId: string;
  title: string;
  content: string;
  tags?: string[];
}

interface CreatePostData {
  threadId: string;
  userId: string;
  content: string;
  parentId?: string;
}

interface ForumSearchParams {
  query?: string;
  categoryId?: string;
  tags?: string[];
  userId?: string;
  page?: number;
  limit?: number;
  sortBy?: 'latest' | 'popular' | 'unanswered';
}

export class ForumService {
  // Get all forum categories
  async getCategories(): Promise<any[]> {
    const cacheKey = 'forum:categories';
    const cached = await getCacheService().get<any[]>(cacheKey);

    if (cached) {
      return cached;
    }

    const categories = await ForumCategory.findAll({
      where: { isActive: true },
      order: [['orderIndex', 'ASC']],
      attributes: ['id', 'name', 'description', 'slug', 'icon', 'color'],
    });

    await getCacheService().set(cacheKey, categories, { ttl: 3600 }); // Cache for 1 hour
    return categories;
  }

  // Create a new thread
  async createThread(data: CreateThreadData): Promise<unknown> {
    const transaction = await sequelize.transaction();

    try {
      // Validate category exists
      const category = await ForumCategory.findByPk(data.categoryId);
      if (!category || !category.isActive) {
        throw new Error('Invalid category');
      }

      // Sanitize content
      const sanitizedContent = this.sanitizeContent(data.content);

      // Create thread
      const thread = await ForumThread.create(
        {
          categoryId: data.categoryId,
          userId: data.userId,
          title: data.title.trim(),
          content: sanitizedContent,
          tags: data.tags || [],
        },
        { transaction }
      );

      // Create initial post
      await ForumPost.create(
        {
          threadId: thread.id,
          userId: data.userId,
          content: sanitizedContent,
        },
        { transaction }
      );

      await transaction.commit();

      // Clear category threads cache
      await getCacheService().del(`forum:threads:${data.categoryId}:*`);

      // Track activity
      await this.trackActivity(data.userId, 'thread_created', 'thread', thread.id);

      return thread;
    } catch (error) {
      await transaction.rollback();
      logger.error('Failed to create thread', { error, data });
      throw error;
    }
  }

  // Get threads with pagination
  async getThreads(params: ForumSearchParams): Promise<{
    threads: unknown[];
    total: number;
    pages: number;
  }> {
    const page = params.page || 1;
    const limit = params.limit || 20;
    const offset = (page - 1) * limit;

    const where: unknown = {};
    const order: unknown[] = [];

    // Filter by category
    if (params.categoryId) {
      where.categoryId = params.categoryId;
    }

    // Filter by tags
    if (params.tags && params.tags.length > 0) {
      where.tags = { [Op.overlap]: params.tags };
    }

    // Filter by user
    if (params.userId) {
      where.userId = params.userId;
    }

    // Search query
    if (params.query) {
      where[Op.or as unknown] = [
        { title: { [Op.iLike]: `%${params.query}%` } },
        { content: { [Op.iLike]: `%${params.query}%` } },
      ];
    }

    // Sorting
    switch (params.sortBy) {
      case 'popular':
        order.push(['views', 'DESC']);
        break;
      case 'unanswered':
        where.replyCount = 0;
        order.push(['createdAt', 'DESC']);
        break;
      case 'latest':
      default:
        order.push(['createdAt', 'DESC']);
    }

    // Add pinned threads first
    order.unshift(['isPinned', 'DESC']);

    const { count, rows: threads } = await ForumThread.findAndCountAll({
      where,
      order,
      limit,
      offset,
      include: [
        {
          model: User,
          as: 'user',
          attributes: ['id', 'name', 'avatar'],
        },
        {
          model: ForumCategory,
          as: 'category',
          attributes: ['id', 'name', 'slug', 'color'],
        },
      ],
    });

    return {
      threads,
      total: count,
      pages: Math.ceil(count / limit),
    };
  }

  // Get thread details with posts
  async getThread(threadId: string, userId?: string): Promise<any | null> {
    const thread = await ForumThread.findByPk(threadId, {
      include: [
        {
          model: User,
          as: 'user',
          attributes: ['id', 'name', 'avatar', 'reputation_points'],
        },
        {
          model: ForumCategory,
          as: 'category',
          attributes: ['id', 'name', 'slug', 'color'],
        },
        {
          model: ForumPost,
          as: 'posts',
          where: { isDeleted: false },
          include: [
            {
              model: User,
              as: 'user',
              attributes: ['id', 'name', 'avatar', 'reputation_points'],
            },
            {
              model: ForumVote,
              as: 'votes',
              attributes: ['userId', 'voteType'],
            },
          ],
          order: [['createdAt', 'ASC']],
        },
      ],
    });

    if (!thread) {
      return null;
    }

    // Increment view count
    await thread.incrementViews();

    // Calculate vote scores for posts
    if (thread.posts) {
      thread.posts.forEach((post: unknown) => {
        const votes = post.votes || [];
        post.voteScore = votes.reduce((sum: number, vote: unknown) => sum + vote.voteType, 0);

        // Add user vote if userId provided
        if (userId) {
          const userVote = votes.find((v: unknown) => v.userId === userId);
          post.userVote = userVote ? userVote.voteType : 0;
        }
      });
    }

    return thread;
  }

  // Create a reply
  async createPost(data: CreatePostData): Promise<unknown> {
    const transaction = await sequelize.transaction();

    try {
      // Validate thread exists and is not locked
      const thread = await ForumThread.findByPk(data.threadId);
      if (!thread) {
        throw new Error('Thread not found');
      }
      if (thread.isLocked) {
        throw new Error('Thread is locked');
      }

      // Validate parent post if provided
      if (data.parentId) {
        const parentPost = await ForumPost.findByPk(data.parentId);
        if (!parentPost || parentPost.threadId !== data.threadId) {
          throw new Error('Invalid parent post');
        }
      }

      // Sanitize content
      const sanitizedContent = this.sanitizeContent(data.content);

      // Create post
      const post = await ForumPost.create(
        {
          threadId: data.threadId,
          userId: data.userId,
          parentId: data.parentId,
          content: sanitizedContent,
        },
        { transaction }
      );

      await transaction.commit();

      // Get post with associations
      const fullPost = await ForumPost.findByPk(post.id, {
        include: [
          {
            model: User,
            as: 'user',
            attributes: ['id', 'name', 'avatar', 'reputation_points'],
          },
        ],
      });

      // Track activity
      await this.trackActivity(data.userId, 'post_created', 'post', post.id);

      // Send notifications
      await this.notifyThreadParticipants(data.threadId, data.userId, post.id);

      return fullPost!;
    } catch (error) {
      await transaction.rollback();
      logger.error('Failed to create post', { error, data });
      throw error;
    }
  }

  // Vote on a post
  async votePost(postId: string, userId: string, voteType: 1 | -1): Promise<number> {
    const transaction = await sequelize.transaction();

    try {
      // Check if vote exists
      const existingVote = await ForumVote.findOne({
        where: { postId, userId },
      });

      if (existingVote) {
        if (existingVote.voteType === voteType) {
          // Remove vote if same type
          await existingVote.destroy({ transaction });
        } else {
          // Update vote type
          existingVote.voteType = voteType;
          await existingVote.save({ transaction });
        }
      } else {
        // Create new vote
        await ForumVote.create(
          {
            postId,
            userId,
            voteType,
          },
          { transaction }
        );
      }

      // Calculate new score
      const votes = await ForumVote.findAll({
        where: { postId },
        attributes: ['voteType'],
        transaction,
      });

      const score = votes.reduce((sum, vote) => sum + vote.voteType, 0);

      await transaction.commit();

      // Update post author reputation
      const post = await ForumPost.findByPk(postId);
      if (post) {
        await this.updateUserReputation(post.userId, voteType * 5);
      }

      return score;
    } catch (error) {
      await transaction.rollback();
      logger.error('Failed to vote on post', { error, postId, userId });
      throw error;
    }
  }

  // Edit a post
  async editPost(postId: string, userId: string, content: string): Promise<unknown> {
    const post = await ForumPost.findByPk(postId);

    if (!post) {
      throw new Error('Post not found');
    }

    if (post.userId !== userId) {
      throw new Error('Unauthorized');
    }

    if (post.isDeleted) {
      throw new Error('Cannot edit deleted post');
    }

    // Sanitize content
    const sanitizedContent = this.sanitizeContent(content);

    // Update post
    post.content = sanitizedContent;
    post.editCount += 1;
    post.lastEditedAt = new Date();
    await post.save();

    return post;
  }

  // Delete a post (soft delete)
  async deletePost(postId: string, userId: string, isAdmin: boolean = false): Promise<void> {
    const post = await ForumPost.findByPk(postId);

    if (!post) {
      throw new Error('Post not found');
    }

    if (!isAdmin && post.userId !== userId) {
      throw new Error('Unauthorized');
    }

    post.isDeleted = true;
    post.deletedAt = new Date();
    post.deletedBy = userId;
    await post.save();
  }

  // Mark post as solution
  async markAsSolution(postId: string, userId: string): Promise<void> {
    const post = await ForumPost.findByPk(postId, {
      include: [
        {
          model: ForumThread,
          as: 'thread',
        },
      ],
    });

    if (!post || !post.thread) {
      throw new Error('Post not found');
    }

    // Only thread author can mark solution
    if (post.thread.userId !== userId) {
      throw new Error('Only thread author can mark solution');
    }

    // Remove previous solution
    await ForumPost.update({ isSolution: false }, { where: { threadId: post.threadId } });

    // Mark new solution
    post.isSolution = true;
    await post.save();

    // Award reputation to solution author
    await this.updateUserReputation(post.userId, 50);
  }

  // Helper methods

  private sanitizeContent(content: string): string {
    if (!content || typeof content !== 'string') {
      return '';
    }

    // Configure DOMPurify for forum content
    const cleanConfig = {
      ALLOWED_TAGS: [
        'p', 'br', 'strong', 'b', 'em', 'i', 'u', 'strike', 's',
        'blockquote', 'code', 'pre', 'ul', 'ol', 'li',
        'h1', 'h2', 'h3', 'h4', 'h5', 'h6',
        'a', 'img'
      ],
      ALLOWED_ATTR: {
        'a': ['href', 'title', 'target', 'rel'],
        'img': ['src', 'alt', 'title', 'width', 'height'],
        '*': ['class']
      },
      ALLOW_DATA_ATTR: false,
      FORBID_TAGS: ['script', 'object', 'embed', 'iframe', 'form', 'input', 'textarea', 'select', 'button'],
      FORBID_ATTR: ['onerror', 'onload', 'onclick', 'onmouseover', 'onfocus', 'onblur', 'style'],
      KEEP_CONTENT: true,
      RETURN_DOM: false,
      RETURN_DOM_FRAGMENT: false,
      SANITIZE_DOM: true,
      SANITIZE_NAMED_PROPS: true,
      WHOLE_DOCUMENT: false,
      // Security-focused configurations
      FORCE_BODY: false,
      RETURN_TRUSTED_TYPE: false,
      SAFE_FOR_TEMPLATES: true
    };

    try {
      // First pass: Basic validation and length check
      if (content.length > 50000) { // 50KB limit for forum posts
        content = content.substring(0, 50000);
      }

      // Second pass: Remove potentially dangerous patterns
      content = content
        .replace(/javascript:/gi, 'blocked:')
        .replace(/data:/gi, 'blocked:')
        .replace(/vbscript:/gi, 'blocked:')
        .replace(/on\w+\s*=/gi, '')
        .replace(/<\s*script[^>]*>.*?<\s*\/\s*script\s*>/gis, '')
        .replace(/<\s*iframe[^>]*>.*?<\s*\/\s*iframe\s*>/gis, '')
        .replace(/<\s*object[^>]*>.*?<\s*\/\s*object\s*>/gis, '')
        .replace(/<\s*embed[^>]*>.*?<\s*\/\s*embed\s*>/gis, '');

      // Third pass: DOMPurify sanitization
      const sanitized = DOMPurify.sanitize(content, cleanConfig);

      // Fourth pass: Additional validation for links
      const linkRegex = /<a[^>]+href\s*=\s*["']([^"']+)["'][^>]*>/gi;
      const processedContent = sanitized.replace(linkRegex, (match, href) => {
        // Validate URL protocols
        const allowedProtocols = ['http:', 'https:', 'mailto:'];
        try {
          const url = new URL(href);
          if (!allowedProtocols.includes(url.protocol)) {
            return match.replace(href, '#blocked-url');
          }
          // Add security attributes to external links
          if (url.hostname !== 'localhost' && !url.hostname.includes('upcoach')) {
            return match.replace('>', ' rel="noopener noreferrer nofollow" target="_blank">');
          }
          return match;
        } catch {
          // Invalid URL - remove the link but keep content
          return match.replace(/<a[^>]*>/gi, '').replace(/<\/a>/gi, '');
        }
      });

      return processedContent.trim();
    } catch (error) {
      logger.error('Content sanitization failed', { error, contentLength: content.length });
      // Fallback to very basic sanitization
      return content
        .replace(/<[^>]*>/g, '') // Strip all HTML tags
        .replace(/[<>"'&]/g, '') // Remove potentially dangerous characters
        .trim()
        .substring(0, 10000); // Limit length
    }
  }

  private async trackActivity(
    userId: string,
    type: string,
    targetType: string,
    targetId: string
  ): Promise<void> {
    try {
      await sequelize.query(
        `INSERT INTO activities (user_id, activity_type, target_type, target_id, created_at)
         VALUES (:userId, :type, :targetType, :targetId, NOW())`,
        {
          replacements: { userId, type, targetType, targetId },
        }
      );
    } catch (error) {
      logger.error('Failed to track activity', { error, userId, type });
    }
  }

  private async updateUserReputation(userId: string, points: number): Promise<void> {
    try {
      await sequelize.query(
        `UPDATE users SET reputation_points = reputation_points + :points WHERE id = :userId`,
        {
          replacements: { userId, points },
        }
      );
    } catch (error) {
      logger.error('Failed to update user reputation', { error, userId, points });
    }
  }

  private async notifyThreadParticipants(
    threadId: string,
    authorId: string,
    postId: string
  ): Promise<void> {
    try {
      // Get all unique participants in thread
      const participants = await sequelize.query(
        `SELECT DISTINCT user_id FROM forum_posts 
         WHERE thread_id = :threadId 
         AND user_id != :authorId 
         AND is_deleted = false`,
        {
          replacements: { threadId, authorId },
          type: 'SELECT',
        }
      );

      // Create notifications for each participant
      const notifications = participants.map((p: unknown) => ({
        user_id: p.user_id,
        type: 'forum_reply',
        title: 'New reply in thread',
        message: 'Someone replied to a thread you participated in',
        data: { threadId, postId },
        created_at: new Date(),
      }));

      if (notifications.length > 0) {
        await sequelize.query(
          `INSERT INTO notifications (user_id, type, title, message, data, created_at)
           VALUES ${notifications.map(() => '(?, ?, ?, ?, ?, ?)').join(', ')}`,
          {
            replacements: notifications.flatMap(n => [
              n.user_id,
              n.type,
              n.title,
              n.message,
              JSON.stringify(n.data),
              n.created_at,
            ]),
          }
        );
      }
    } catch (error) {
      logger.error('Failed to notify thread participants', { error, threadId });
    }
  }
}

export const forumService = new ForumService();

File: apps/mobile/lib/features/community/services/forum_service.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../../core/services/api_service.dart';
import '../presentation/providers/forum_providers.dart';

class ForumService {
  final ApiService _apiService;

  ForumService(this._apiService);

  /// Get all forum categories
  Future<List<ForumCategory>> getCategories() async {
    final response = await _apiService.get('/forum/categories');

    if (response.data['success'] == true) {
      final List<dynamic> categoriesJson = response.data['data'] ?? [];
      return categoriesJson
          .map((json) => ForumCategory(
                id: json['id'] as String,
                name: json['name'] as String,
                description: json['description'] as String? ?? '',
                icon: json['icon'] as String? ?? 'chat',
                threadCount: json['threadCount'] as int? ?? 0,
                memberCount: json['memberCount'] as int? ?? 0,
              ))
          .toList();
    }

    throw Exception(response.data['error'] ?? 'Failed to load categories');
  }

  /// Get threads with optional filtering
  Future<ThreadsResponse> getThreads({
    String? categoryId,
    List<String>? tags,
    String? userId,
    String? query,
    int page = 1,
    int limit = 20,
    String sortBy = 'latest',
  }) async {
    final queryParams = <String, dynamic>{
      'page': page.toString(),
      'limit': limit.toString(),
      'sortBy': sortBy,
    };

    if (categoryId != null) queryParams['categoryId'] = categoryId;
    if (tags != null && tags.isNotEmpty) queryParams['tags'] = tags.join(',');
    if (userId != null) queryParams['userId'] = userId;
    if (query != null && query.isNotEmpty) queryParams['query'] = query;

    final response = await _apiService.get(
      '/forum/threads',
      queryParameters: queryParams,
    );

    if (response.data['success'] == true) {
      final List<dynamic> threadsJson = response.data['data'] ?? [];
      final pagination = response.data['pagination'] as Map<String, dynamic>?;

      final threads = threadsJson.map((json) => _parseThread(json)).toList();

      return ThreadsResponse(
        threads: threads,
        total: pagination?['total'] as int? ?? threads.length,
        pages: pagination?['pages'] as int? ?? 1,
        page: pagination?['page'] as int? ?? page,
        limit: pagination?['limit'] as int? ?? limit,
      );
    }

    throw Exception(response.data['error'] ?? 'Failed to load threads');
  }

  /// Get a single thread by ID
  Future<ForumThread?> getThread(String threadId) async {
    final response = await _apiService.get('/forum/threads/$threadId');

    if (response.data['success'] == true) {
      final threadJson = response.data['data'];
      if (threadJson != null) {
        return _parseThread(threadJson);
      }
    }

    return null;
  }

  /// Create a new thread
  Future<ForumThread> createThread({
    required String categoryId,
    required String title,
    required String content,
    List<String>? tags,
  }) async {
    final response = await _apiService.post(
      '/forum/threads',
      data: {
        'categoryId': categoryId,
        'title': title,
        'content': content,
        if (tags != null && tags.isNotEmpty) 'tags': tags,
      },
    );

    if (response.data['success'] == true) {
      return _parseThread(response.data['data']);
    }

    throw Exception(response.data['error'] ?? 'Failed to create thread');
  }

  /// Create a post/reply in a thread
  Future<ForumPost> createPost({
    required String threadId,
    required String content,
    String? parentId,
  }) async {
    final response = await _apiService.post(
      '/forum/posts',
      data: {
        'threadId': threadId,
        'content': content,
        if (parentId != null) 'parentId': parentId,
      },
    );

    if (response.data['success'] == true) {
      return _parsePost(response.data['data']);
    }

    throw Exception(response.data['error'] ?? 'Failed to create post');
  }

  /// Vote on a post
  Future<int> votePost(String postId, int voteType) async {
    final response = await _apiService.post(
      '/forum/posts/$postId/vote',
      data: {'voteType': voteType},
    );

    if (response.data['success'] == true) {
      return response.data['data']['score'] as int? ?? 0;
    }

    throw Exception(response.data['error'] ?? 'Failed to vote');
  }

  /// Edit a post
  Future<ForumPost> editPost(String postId, String content) async {
    final response = await _apiService.patch(
      '/forum/posts/$postId',
      data: {'content': content},
    );

    if (response.data['success'] == true) {
      return _parsePost(response.data['data']);
    }

    throw Exception(response.data['error'] ?? 'Failed to edit post');
  }

  /// Delete a post
  Future<void> deletePost(String postId) async {
    final response = await _apiService.delete('/forum/posts/$postId');

    if (response.data['success'] != true) {
      throw Exception(response.data['error'] ?? 'Failed to delete post');
    }
  }

  /// Mark a post as the solution
  Future<void> markAsSolution(String postId) async {
    final response = await _apiService.post('/forum/posts/$postId/solution');

    if (response.data['success'] != true) {
      throw Exception(response.data['error'] ?? 'Failed to mark as solution');
    }
  }

  /// Get activity feed
  Future<List<ActivityFeedItem>> getActivityFeed({int limit = 50}) async {
    final response = await _apiService.get(
      '/forum/activity',
      queryParameters: {'limit': limit.toString()},
    );

    if (response.data['success'] == true) {
      final List<dynamic> feedJson = response.data['data'] ?? [];
      return feedJson
          .map((json) => ActivityFeedItem(
                id: json['id'] as String,
                type: json['type'] as String,
                userId: json['userId'] as String,
                userName: json['userName'] as String? ?? 'Unknown',
                userAvatar: json['userAvatar'] as String?,
                content: json['content'] as String? ?? '',
                createdAt: DateTime.parse(
                    json['createdAt'] as String? ?? DateTime.now().toIso8601String()),
                metadata: json['metadata'] as Map<String, dynamic>?,
              ))
          .toList();
    }

    throw Exception(response.data['error'] ?? 'Failed to load activity feed');
  }

  /// Get community groups
  Future<List<CommunityGroup>> getCommunityGroups() async {
    final response = await _apiService.get('/community/groups');

    if (response.data['success'] == true) {
      final List<dynamic> groupsJson = response.data['data'] ?? [];
      return groupsJson
          .map((json) => CommunityGroup(
                id: json['id'] as String,
                name: json['name'] as String,
                description: json['description'] as String? ?? '',
                imageUrl: json['imageUrl'] as String?,
                avatarUrl: json['avatarUrl'] as String?,
                memberCount: json['memberCount'] as int? ?? 0,
                postCount: json['postCount'] as int? ?? 0,
                isPrivate: json['isPrivate'] as bool? ?? false,
                createdBy: json['createdBy'] as String? ?? '',
                createdAt: DateTime.parse(
                    json['createdAt'] as String? ?? DateTime.now().toIso8601String()),
              ))
          .toList();
    }

    throw Exception(response.data['error'] ?? 'Failed to load community groups');
  }

  ForumThread _parseThread(Map<String, dynamic> json) {
    return ForumThread(
      id: json['id'] as String,
      categoryId: json['categoryId'] as String,
      title: json['title'] as String,
      content: json['content'] as String? ?? '',
      authorId: json['authorId'] as String? ?? json['userId'] as String? ?? '',
      authorName: json['authorName'] as String? ??
          json['author']?['name'] as String? ??
          'Unknown',
      authorAvatar:
          json['authorAvatar'] as String? ?? json['author']?['avatar'] as String?,
      createdAt: DateTime.parse(
          json['createdAt'] as String? ?? DateTime.now().toIso8601String()),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
      replyCount: json['replyCount'] as int? ?? json['postCount'] as int? ?? 0,
      viewCount: json['viewCount'] as int? ?? 0,
      isPinned: json['isPinned'] as bool? ?? false,
      isLocked: json['isLocked'] as bool? ?? false,
      tags: (json['tags'] as List<dynamic>?)?.cast<String>() ?? [],
    );
  }

  ForumPost _parsePost(Map<String, dynamic> json) {
    return ForumPost(
      id: json['id'] as String,
      threadId: json['threadId'] as String,
      content: json['content'] as String,
      authorId: json['authorId'] as String? ?? json['userId'] as String? ?? '',
      authorName: json['authorName'] as String? ??
          json['author']?['name'] as String? ??
          'Unknown',
      authorAvatar:
          json['authorAvatar'] as String? ?? json['author']?['avatar'] as String?,
      createdAt: DateTime.parse(
          json['createdAt'] as String? ?? DateTime.now().toIso8601String()),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
      score: json['score'] as int? ?? 0,
      isSolution: json['isSolution'] as bool? ?? false,
      parentId: json['parentId'] as String?,
    );
  }
}

/// Forum post model
class ForumPost {
  final String id;
  final String threadId;
  final String content;
  final String authorId;
  final String authorName;
  final String? authorAvatar;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final int score;
  final bool isSolution;
  final String? parentId;

  const ForumPost({
    required this.id,
    required this.threadId,
    required this.content,
    required this.authorId,
    required this.authorName,
    this.authorAvatar,
    required this.createdAt,
    this.updatedAt,
    this.score = 0,
    this.isSolution = false,
    this.parentId,
  });
}

/// Response model for paginated threads
class ThreadsResponse {
  final List<ForumThread> threads;
  final int total;
  final int pages;
  final int page;
  final int limit;

  const ThreadsResponse({
    required this.threads,
    required this.total,
    required this.pages,
    required this.page,
    required this.limit,
  });
}

/// Provider for ForumService
final forumServiceProvider = Provider<ForumService>((ref) {
  final apiService = ref.watch(apiServiceProvider);
  return ForumService(apiService);
});
